### PR TITLE
feat(): Breaking restructure

### PR DESCRIPTION
**Pull request by kaosat-dev**
WORK IN PROGRESS, DO NOT MERGE !
This is the first' break all the old stuff' PR for the V2 branch !
It 
- [x] removes most aliases
- [x] removes most cyclic dependencies
- [x] restructures a number of parts of the code to be api-centric

### CODE REVIEW

**Comment by z3dev** [comment]
difference.js... really don’t like this feature...
``

File: src/api/ops-booleans/difference.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+function difference () {
+  let object
+  let i = 0
+  let a = arguments
+  if (a[0].length) a = a[0]
+  for (object = a[i++]; i < a.length; i++) {
+    if (isCAG(a[i])) {
+      object = object.subtract(a[i])
+    } else {
+      object = object.subtract(a[i].setColor(1, 1, 0)) // -- color the cuts
```
Comment (z3dev) [inline src/api/ops-booleans/difference.js]: Can we remove this ‘feature’? It’s really annoying when doing complex designs which require colors.

Comment (kaosat-dev) [reply src/api/ops-booleans/difference.js]: By feature dobyou mean just the part that colors the difference ?

**Comment by z3dev** [comment]
Really need to decide the order of the parameters.... there’s a lot of mixing happening.

Function(objects, options)

Function(options, objects)

**Comment by kaosat-dev** [comment]
I am systematically changing everything to (options, objects).That ordering is for a reason btw : it allows us to create curryable functions, to use the (...objects) syntax etc

**Comment by z3dev** [comment]
function (options, objects).

I hope bought that this was the syntax required... :)

**Comment by kaosat-dev** [comment]
Yep, always options first :-)

File: src/api/ops-booleans/union.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+function union () {
+  let options = {}
+  const defaults = {
+    extrude2d: false
+  }
+  let o
+  let i = 0
+  let a = arguments
+  if (a[0].length) a = a[0]
+  if ('extrude2d' in a[0]) { // first parameter is options
```
[the remark this answers is not in the document]
Comment (kaosat-dev) [reply src/api/ops-booleans/union.js]: do you mean the implicit extrusion ? yes that will go away

File: src/api/ops-extrusions/extrusionUtils.js [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+// THIS IS AN OLD untested !!! version of rotate extrude
+/** Extrude to into a 3D solid by rotating the origin around the Y axis.
+ * (and turning everything into XY plane)
+ * @param {Object} options - options for construction
+ * @param {Number} [options.angle=360] - angle of rotation
+ * @param {Number} [options.resolution=defaultResolution3D] - number of polygons per 360 degree revolution
+ * @returns {CSG} new 3D solid
+ */
+const rotateExtrude = function (cag, options) { // FIXME options should be optional
```
[the remark this answers is not in the document]
Comment (kaosat-dev) [reply src/api/ops-extrusions/extrusionUtils.js]: I am far from done with the cleanup :) but yes, only one shall remain !

File: src/api/ops-booleans/union.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (z3dev) [reply src/api/ops-booleans/union.js]: Correct. No implicit anything.

**Comment by z3dev** [comment]
I consider compactBinary()should be part of IO.

**Comment by z3dev** [comment]
Hmmm... maybe JSON should be part of the CSG library? The difference between compact and JSON is trivial, as both are alternates to javascript object serialization. Hmmm....

**Comment by z3dev** [comment]
Can this PR be renamed to V2API ...?

**Comment by z3dev** [comment]
ops-booleans.js

Good, but I think we should restrict the parameters to a strict pattern, I.e. function (options, objects)

Options is REQUIRED, or detectable by testing the first argument.

Also, this check is common to all functions.

Finally, the application of the operations should be based on the first CSG/ CAG/etc. Normal procession continues without issues.

What should processing do if given MIXED object types? I think FAIL. Alternatives are... CONTINUE or BAILOUT.

**Comment by kaosat-dev** [comment]
err , did you check out the code @z3dev  ? I already added detection and throwing an exception if you pass mixed types, and I also added the test for that :) 
Parameter orders are now changed to always be `options, objects` except for booleans where there are no options to be passed :)

**Comment by kaosat-dev** [comment]
I understand the logic of `compactBinary` being possibly part of IO, except it is a very low level conversion, and needed by the 'core' and IO can depend on core, but not the other way around :)
for JSON : I do not know yet : the 'virtual tree' system already manipulates plain JS objects (that are almost JSON), so for the final JSON format I think we need to wait a bit for things to 'settle' :)

**Comment by z3dev** [comment]
Colors...
- need to decide 0-255 or 0.0 to 1.0 (colors in WebGL go from 0 to 1, but possibly decided by the GL library requirements)
- hue2rgb() ?? no idea what this function does... any ideas?
- parameters to functions need to be consistent with general API

**Comment by z3dev** [comment]
Looking at the progress...
- glad to see path2 moved into core, from math. it belongs here.
- please rename Line2.js and Line3.js to Ray2.js and Ray3.js, this will make room for real lines later.
- getTag()... just thinking but could there be problems with multiple threads calling this routine? obtaining the same number?
- toString() methods could use ES6 string templates ;)
- api, ops-transformations/mirror.js, line 23 says 'scale' ??
- core/geometry/shape*/*X.js *Y.js *Z.js ... if these are staying then move to API, core should only have matrix based functions for transforms

By the way, once you get to a fairly solid code base, with the restructure complete, then I can contribute directly with some changes / fixes. It still seems a little early. GANBATTE

**Comment by z3dev** [comment]
Just a thought... but with the move to a functional API, we should use different verbs for get* functions. I think the use of get implies an object. Some options are...

call, conjecture, estimate, figure, gauge (also gage), guess, judge, make, place, put, reckon, suppose
conclude, deduce, extrapolate, gather, infer, reason, understand
calibrate, measure, scale
compute, work out

It should probably infer some kind of work, like compute. There are now stored values for area, center, etc.

**Comment by kaosat-dev** [comment]
@z3dev going to merge this into the V2.0 branch despite it not being final, so you can add the changes on top ok ?

**Pull request merged**